### ISSUE TITLE
Implement download cache

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#281]
When creating new package recipes, I often have to use `spack clean` to remove left-over previous partial builds. This also deletes the downloaded archives, and re-downloading them takes time. As the downloaded files have checksums, this shouldn't be necessary, and a general download cache (outside of the staging directories etc.) would make sense.

**Comment by nrichart** [comment]
According to #277, I guess that they are already thinking on ways to change this. But in the mean time you can give a look at `spack mirror`. It will still fetch the files to the stage directory, but instead of downloading it is only a copy the mirrored files.

**Comment by citibeth** [comment]
Is it possible to add packages to a Spack mirror after the mirror has been created?

**Comment by eschnett** [comment]
Yes, this works. I think the default mirror name is the date, so this works best if you add a package on the same day. Of course, you will still have to manually copy the new mirror to the new system.

**Comment by tgamblin** [comment]
`spack mirror` commands also let you specify a name for the mirror, which doesn't have to include the date, but yes the default includes the date.  If people have ideas for better default mirror names, I'd be interested.

**Comment by citibeth** [comment]
I think the problem with the mirror is that it downloads EVERYTHING, which
is just too much.  Better would be a way to download a subset of tarballs.
For example... everything that's currently installed in this Spack, or
everything needed by a set of specs.

**Comment by tgamblin** [comment]
You can do that by either supplying a file or extra arguments to the mirror command.  If you want fewer tarballs, you can be more specific:

```
$ spack mirror create -h
usage: spack mirror create [-h] [-d DIRECTORY] [-f FILE] [-D] [-o] ...

positional arguments:
  specs                 Specs of packages to put in mirror

optional arguments:
  -h, --help            show this help message and exit
  -d DIRECTORY, --directory DIRECTORY
                        Directory in which to create mirror.
  -f FILE, --file FILE  File with specs of packages to put in mirror.
  -D, --dependencies    Also fetch all dependencies
  -o, --one-version-per-spec
                        Only fetch one 'preferred' version per spec, not all
                        known versions.
```

e.g.:

```
spack mirror create -d mymirror -f list-of-specs.txt
```

or

```
spack mirror create -d mymirror foo@1.0.1 bar@1.0.2 [...]
```

etc.

**Comment by citibeth** [comment]
Thanks. I didn't pick this up from the docs.

**Comment by scheibelp** [comment]
FWIW I started a PR on this: https://github.com/LLNL/spack/pull/579

It creates a var/spack/cache directory and treats it as a default mirror for downloads (i.e. one that doesn't have to be specified in the config); packages are added to the mirror as they are downloaded.

**Comment by citibeth** [comment]
Looks like good work!

> This does not deal with caching for source control repositories
> (i.e. does not modify VCSFetchStrategy).

I'm wondering where we want to go with this.  Tickled with the right URL,
GitHub and Bitbucket will give you a tarball for any version they have in
the repo.  So far, I haven't had to use VCSFetchStrategy, which would seem
to be less efficient (unless you do git clone --depth 1, not sure if that
works on all versions of git).

For now, I would not bother implementing caching for VCSFetchStrategy.
Maybe we'll never find that lack to be a serious problem.

On a related topic... if you download via VCSFetchStrategy or a Git
tarball, Autotools-based projects will not usually provide a ./configure
script.  I've found this "problem" is easily resolved by:
  1. Add depends_on('autoconf') to the package.
  2. Add an autoreconf call to the install() method, eg:
```
        which('autoreconf')('-i')
        configure(*configure_args)
        make()
        make("install")
```

I'm wondering whether that should be standard for all Autotools-based
packages.

-- Elizabeth

**Comment by scheibelp** [comment]
The PR has changed since the time of the initial comment so that it should support VCSFetchingStrategy. I apologize for not clarifying that anywhere and I just added an edit to the initial comment. 

Todd made some suggestions (which github collapses since they reference code which has changed), for example to use FetchStrategy.archive which (for the VCS implementations) creates a tar.gz identified by the commit/branch.

**Comment by adamjstewart** [comment]
Is this issue resolved now that #579 has been merged?

**Comment by becker33** [comment]
Yes, appears so.
